src/main/scala/is/hail/expr/FunctionRegistry.scala

```diff
@@ -237,10 +226,109 @@ object FunctionRegistry {
           v = args(2).run(ec)();
 
           _ = (if (v == null)
-            fatal(s"Argument evaluated to missing in call to aggregator $name"));
+            fatal(s"Argument evaluated to missing in call to aggregator $name"))
+
+        ) yield aggregator.ctor(g, v.asInstanceOf[v])
 
-          aggregationResultThunk <- addAggregation(args(0), aggregator.ctor(g, v.asInstanceOf[v]));
+      case aggregator: GroupByAggregatorFun[t] =>
+        val Lambda(_, keyParam, keyBody) = args(1)
+        val TFunction(Seq(keyParamType), _) = argTypes(1)
 
+        val Lambda(_, downstreamParam, downstreamBody) = args(2)
+        val TFunction(Seq(downstreamParamType), _) = argTypes(2)
+
+        val kbc = keyBody.compile()
+
+        val a0type = args(0).`type`
+
+        for (
+          (downstreamBodyCompiled, downstreamAggregator) <- downstreamBody.downstreamAggregator();
+
+          ec <- ec();
+          (idx, localA) <- ecNewPosition();
+
+          bodyST = a0type match {
+            case tagg: TAggregable => tagg.symTab
+            case _ => ec.st
+          };
+
+          keyBodyFn = (for (
+            fb <- fb();
+            bindings = (bodyST.toSeq
+              .map { case (name, (i, typ)) =>
+                (name, typ, ret(Code.checkcast(fb.arg2.invoke[Int, AnyRef]("apply", i))(typ.scalaClassTag)))
+            } :+ ((keyParam, keyParamType, ret(Code.checkcast(fb.arg2.invoke[Int, AnyRef]("apply", idx))(keyParamType.scalaClassTag)))));
+            res <- bindRepInRaw(bindings)(kbc)
+          ) yield res).runWithDelayedValues(bodyST.toSeq.map { case (name, (_, typ)) => (name, typ) }, ec);
+
+          key = (x: Any) => {
+            localA(idx) = x
+            keyBodyFn(localA.asInstanceOf[mutable.ArrayBuffer[AnyRef]])
+          };
+
+          (downstreamParamId, localA2) <- ecNewPosition();
+
+          downstreamParamTypeFixed = {
+            val x = downstreamParamType.asInstanceOf[TAggregable].copy();
+            // FIXME filter out old bindings
```

Comment:
@tpoterba You commented on the other PR:
> does this work?
I think it does, but for terrible reasons. I'm really beginning to think Aggregators need to be rethought because I can barely keep my head around them.

So how bindings normally work is that you have your normal EvalContext which for variant parallel stuff contains things like `global`, `v`, `va`, and `gs`. The `gs` binding is special: it has an internal symbol table that *must* (although this isn't completely enforced) overlap exactly with the outer table, but may contain extra bindings. For variant aggregations, this means it has `global`, `v`, `va`, and `g` (which occupies the same position as `gs`), but it additionally has `s` and `sa`.

So, this looks more or less ok, but I think I may have fucked things up and/or made it way more complicated when I changed aggregators to be in "continuation passing style". I intended to simplify things, but I think I made it worse. For example, what happens when folks use a lambda with their own custom name for `g`?

```scala
vds.annotateVariantsExpr('gs.map(cake => cake.gt).count()')
```

In this case we have to add a binding for `cake` that points to wherever `g` is. I'm not sure what we were doing before my change. After my change, every transformation (i.e. `map`) takes the value of `g` (to give to the lambda) and takes a lambda `Any => Unit` which is the "continuation". It's whatever should happen after the `map` function runs. In this case, that means it just calls `seqOp` on the count aggregator.

Here, this "take the value of `g`" and give it to lambda is done by `keyFn`. It set's a position in the eval context's array buffer to its argument. The `keyBodyFn` knows that `keyParam` (which in this example is `cake`) is at position `idx`.

So this seems complex and not exactly right, but it works.

Now, I think with `groupBy`'s nested aggregators things get even more complex and maybe not even right any more. Here we basically want the exact same *outer* symbol table, but now `gs` should be some new binding. For example:

```scala
vds.annotateVariantsExpr('gs.groupBy(g => sa.population, newGs => newGs.count())')
```

here we want the bindings of the second lambda to still include `global`, `va`, and `v`, but not include `gs` and do include `newGs`.

I think you're right that I've got it wrong, but I need to think through this more to get it right. This is really pushing me towards a clean slate implementation of expr, because I feel like the complexity is getting ahead of me now.